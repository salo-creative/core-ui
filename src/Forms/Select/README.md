# Select

This is a Select component that can be used as a `<select>` html element. It is designed to wrap over child `<option>` tags.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
Import Select from '@salo/core-ui/Forms/Select';
```

Implement as follows

```javascript
<Select>
  <option>Option 1</option>
  <option>Option 2</option>
  <option>Option 3</option>
</Select>
```