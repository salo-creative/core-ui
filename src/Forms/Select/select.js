import React from 'react';
import PropTypes from 'prop-types';

// COMPONENTS & STYLES
import { SelectWrapper, Selector, IconAfter, OuterSelectWrapper } from './select.styles';
import Label from '../components/Label';
import ErrorText from '../components/ErrorText';
import HelperText from '../components/HelperText';

// HELPERS
import { colours } from '../../helpers/colours';

const Select = (props) => {
  const {
    background,
    border,
    children,
    className,
    disabled,
    error,
    errorMessage,
    helperText,
    iconAfter,
    label,
    margin,
    name,
    onBlur,
    onChange,
    required,
    size,
    value
  } = props;

  const height = () => {
    switch (size) {
      case 'M':
        return '4rem';
      case 'L':
      default:
        return '4.5rem';
    }
  };

  return (
    <OuterSelectWrapper margin={ margin } className={ `salo-select ${ className }` }>
      <Label
        className='salo-select__label'
        error={ error }
        label={ label }
        name={ name }
        required={ required }
        size={ size }
      />
      <SelectWrapper
        background={ background }
        error={ error }
        className='salo-select__wrapper'
      >
        <Selector
          aria-required={ required.toString() }
          aria-invalid={ error.toString() }
          border={ border.toString() }
          className='salo-select__field'
          disabled={ disabled }
          error={ error }
          height={ height() }
          id={ name }
          name={ name }
          onBlur={ e => onBlur({ e, value: e.target.value }) }
          onChange={ (e) => onChange({ e, value: e.target.value }) }
          value={ value }
        >
          { children }
        </Selector>
        { iconAfter && (
          <IconAfter error={ error }>
            <svg style={ { width: '24px', height: '24px' } } viewBox='0 0 24 24'>
              <path fill={ colours.darkGrey } d='M7,10L12,15L17,10H7Z' />
            </svg>
          </IconAfter>
        ) }
      </SelectWrapper>
      <ErrorText
        className='salo-select__error'
        disabled={ disabled }
        error={ error }
        errorMessage={ errorMessage }
        size={ size }
      />
      <HelperText
        className='salo-select__helper'
        disabled={ disabled }
        error={ error }
        helperText={ helperText }
        size={ size }
      />
    </OuterSelectWrapper>
  );
};

Select.propTypes = {
  children: PropTypes.array,
  disabled: PropTypes.bool,
  iconAfter: PropTypes.bool,
  onChange: PropTypes.func,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

Select.defaultProps = {
  background: colours.paleGrey,
  border: true,
  children: null,
  className: '',
  disabled: false,
  error: false,
  errorMessage: 'Field invalid',
  helperText: '',
  iconAfter: true,
  label: '',
  margin: '0 0 2rem',
  onBlur: () => null,
  onChange: null,
  required: false,
  size: 'M',
  value: ''
};

Select.propTypes = {
  background: PropTypes.string,
  border: PropTypes.bool,
  children: PropTypes.array,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  helperText: PropTypes.string,
  iconAfter: PropTypes.bool,
  label: PropTypes.string,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M']),
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

export default Select;