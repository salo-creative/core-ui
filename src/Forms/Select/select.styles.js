import styled from 'styled-components';
import { transparentize } from 'polished';

export const SelectWrapper = styled.div`
  position: relative;
  display: inline-block;
  vertical-align: middle;
  margin: 0;
  width: 100%;
  background: ${ ({ background }) => background };
`;

export const OuterSelectWrapper = styled.div`
  width: 100%;
  flex-direction: column;
  display: flex;
  margin: ${ ({ margin }) => margin };
`;

export const Selector = styled.select`
  width: 100%;
  height: ${ ({ height }) => height };
  background: transparent;
  color: ${ ({ theme }) => theme.font };
  font-size: inherit;
  padding: 0 4rem 0 1rem;
  border: 1px solid ${ ({ theme }) => transparentize(0.5, theme.primary) };
  margin: 0;
  border-radius: 0;
  text-overflow: '';
  appearance: none;
  outline: none !important;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  transition: border 0.2s linear;
  position: relative;
  z-index: 2;
  &:focus,
  &:active {
    border-color: ${ ({ theme }) => theme.primary };
  }
  &[disabled] {
    cursor: not-allowed;
    opacity: 0.75;
  }
  ${ ({ border }) => (border === 'false' ? `
    border-color: transparent !important;
    padding: 0 4rem 0 0;
  ` : '') }
`;

export const IconAfter = styled.span`
  display: flex;
  width: 4rem;
  height: calc(100% - 1.5rem);
  align-items: center;
  right: 0;
  top: 0.75rem;
  position: absolute;
  justify-content: center;
  border-left: 1px solid ${ ({ theme }) => transparentize(0.5, theme.primary) };
  z-index: 1;
`;

// TITLE PICKER
export const TitleWrapper = styled.div`
  width: 100%;
  position: relative;
  &.loading,
  &.error {
    &:after {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      content: '';
      background: rgba(255,255,255,0.9);
    }
  }
`;

export const RetryWrapper = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  z-index: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 1.4rem;
`;

export const Text = styled.span`
  margin-left: 1rem;
`;