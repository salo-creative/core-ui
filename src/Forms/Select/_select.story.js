import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, select } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// Tests.
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// Featured component.
import { Select } from '../../index';

// README
import README from './README.md';

// Story logic.
const stories = storiesOf('Forms | Select', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['select'] });

stories.add(
  'Basic',
  (() => {
    // Store.
    const store = new Store({ value: 'Option 1' });
    // Knobs.
    const label = text('Label', 'Select Label');
    const helperText = text('Helper text', 'Helper text');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const iconAfter = boolean('Icon after', true);
    const size = select('size', ['L', 'M'], 'M');
    return (
      <RenderWithProps store={ store }>
        <Select
          name='story'
          disabled={ disabled }
          error={ error }
          label={ label }
          required={ required }
          helperText={ helperText }
          iconAfter={ iconAfter }
          size={ size }
          value={ store.value }
          onChange={ ({ value }) => store.set({ value }) }
        >
          <option value='Option 1'>Option 1</option>
          <option value='Option 2'>Option 2</option>
          <option value='Option 3'>Option 3</option>
        </Select>
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);