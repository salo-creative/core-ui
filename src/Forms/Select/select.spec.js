import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Select from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Select />
  );
  expect(container.firstChild).toMatchSnapshot();
});