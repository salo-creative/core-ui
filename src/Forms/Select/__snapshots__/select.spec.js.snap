// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c1 {
  position: relative;
  display: inline-block;
  vertical-align: middle;
  margin: 0;
  width: 100%;
  background: #fafcff;
}

.c0 {
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 0 2rem;
}

.c2 {
  width: 100%;
  height: 4rem;
  background: transparent;
  color: #333333;
  font-size: inherit;
  padding: 0 4rem 0 1rem;
  border: 1px solid;
  border-color: rgba(0,172,237,0.5);
  margin: 0;
  border-radius: 0;
  text-overflow: '';
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: none !important;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  -webkit-transition: border 0.2s linear;
  transition: border 0.2s linear;
  position: relative;
  z-index: 2;
}

.c2:focus,
.c2:active {
  border-color: #00ACED;
}

.c2[disabled] {
  cursor: not-allowed;
  opacity: 0.75;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 4rem;
  height: calc(100% - 1.5rem);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  right: 0;
  top: 0.75rem;
  position: absolute;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-left: 1px solid;
  border-color: rgba(0,172,237,0.5);
  z-index: 1;
}

<div
  class="c0 salo-select "
>
  <div
    class="c1 salo-select__wrapper"
  >
    <select
      aria-invalid="false"
      aria-required="false"
      class="c2 salo-select__field"
      height="4rem"
    />
    <span
      class="c3"
    >
      <svg
        style="width: 24px; height: 24px;"
        viewBox="0 0 24 24"
      >
        <path
          d="M7,10L12,15L17,10H7Z"
          fill="#6c757d"
        />
      </svg>
    </span>
  </div>
</div>
`;
